c++/src/kj/compat/http.c++

```diff
@@ -1729,33 +1826,49 @@ static_assert(!fastCaseCmp<'f','O','o','B'>("FooB1"), "");
 static_assert(!fastCaseCmp<'f','O','o','B','1','a'>("FooB1"), "");
 
 kj::Own<kj::AsyncInputStream> HttpInputStreamImpl::getEntityBody(
-    RequestOrResponse type, HttpMethod method, uint statusCode,
+    RequestOrResponse type,
+    kj::OneOf<HttpMethod, HttpConnectMethod> method,
+    uint statusCode,
     const kj::HttpHeaders& headers) {
   KJ_REQUIRE(headerBuffer.size() > 0, "Cannot get entity body after header buffer release.");
 
+  auto isHeadRequest = method.tryGet<HttpMethod>().map([](auto& m) {
+    return m == HttpMethod::HEAD;
+  }).orDefault(false);
+
+  auto isConnectRequest = method.is<HttpConnectMethod>();
+
   // Rules to determine how HTTP entity-body is delimited:
   //   https://tools.ietf.org/html/rfc7230#section-3.3.3
-
   // #1
   if (type == RESPONSE) {
-    if (method == HttpMethod::HEAD) {
+    if (isHeadRequest) {
       // Body elided.
       kj::Maybe<uint64_t> length;
       KJ_IF_MAYBE(cl, headers.get(HttpHeaderId::CONTENT_LENGTH)) {
         length = strtoull(cl->cStr(), nullptr, 10);
       } else if (headers.get(HttpHeaderId::TRANSFER_ENCODING) == nullptr) {
-        // HACK: Neither Content-Length nor Transfer-Encoding header in response to HEAD request.
-        //   Propagate this fact with a 0 expected body length.
+        // HACK: Neither Content-Length nor Transfer-Encoding header in response to HEAD
+        //   request. Propagate this fact with a 0 expected body length.
         length = uint64_t(0);
       }
       return kj::heap<HttpNullEntityReader>(*this, length);
+    } else if (isConnectRequest && statusCode >= 200 && statusCode < 300) {
+      // #2 For any 2xx response to a CONNECT request... a client MUST ignore any
+      // Content-Length or Transfer-Encoding header fields received.
+      return kj::heap<HttpConnectionCloseEntityReader>(*this);
```

Comment:
I think the way things are written, we never take this branch, because `HttpClientImpl` will never call `getEntityBody()` for CONNECT responses with 2xx status codes. But if we did take this branch, then returning `HttpConnectionCloseEntityReader` seems wrong, since this interferes with constructing an actual CONNECT stream. Maybe we should just assert-fail here?